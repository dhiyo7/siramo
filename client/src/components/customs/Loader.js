import React, { Component } from 'react'
import {
  StyleSheet,
  View,
  Modal,
  ActivityIndicator
} from 'react-native'
import * as Progress from 'react-native-progress'

const Loader = props => {
  const {
    loading,
    ...attributes
  } = props

  return (
    <Modal 
      visible={loading}
      animationType={'none'}
      transparent={true}
      onRequestClose={() => {console.log('close modal')}}
    >
      <View style={styles.modalBackGround}>
        <View style={styles.ActivityIndicatorWrapper}>
          {/* <ActivityIndicator
          
            animating={loading}
          /> */}
          <Progress.CircleSnail color={['#90CAF9', '#FFAB91', '#A5D6A7']} size={100} thickness={4} spinDuration={10000}/>
        </View>
      </View>
    </Modal>
  )
}

const styles = StyleSheet.create({
  modalBackGround: {
    flex: 1,
    alignItems: 'center',
    flexDirection: 'column',
    justifyContent: 'space-around',
    backgroundColor: '#00000040'
  },
  ActivityIndicatorWrapper: {
    borderRadius: 10,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-around'
  }
})

export default Loader