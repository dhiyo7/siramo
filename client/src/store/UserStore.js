import { observable, computed } from 'mobx'
import { db, User } from './firebase'
import { Alert } from 'react-native'
import firebase from 'firebase'

class UserStore {
  @observable userData = {
    uid: '',
    email: '',
    loading: false
  }

  @observable farmData = {
    name: '',
    temperature: 0,
    water_ratio: 0,
    water_level: 0,
    humidity: 0,
    last_siram: 0,
    last_updated: 0,
    ready_siram: 2,
    loading: false
  }

  @observable isLogin = false

  getFarmData = () => {
    let userId = this.userData.uid
    this.farmData.loading = true
    db.ref(`/farms/${userId}`).on('value', snap => {
      try {
        let farmDB = snap.val()
        console.log('Farm Data ===>', farmDB.last_siram)
        this.farmData.name = farmDB.name
        this.farmData.temperature = farmDB.temperature
        this.farmData.water_ratio = farmDB.water_ratio
        this.farmData.water_level = farmDB.water_level
        this.farmData.humidity = farmDB.humidity
        this.farmData.last_siram = farmDB.last_siram
        this.farmData.last_updated = farmDB.last_updated
        this.farmData.ready_siram = farmDB.ready_siram
        this.farmData.loading = false
      } catch (error) {
        console.log(error)
      }
      
    })
  }

  assignUserData = (data) => {
    this.userData.uid = data.uid
    this.userData.email = data.email
  }

  firebaseLogin = (data, navigation) => {
    this.userData.loading = true
    User.signInWithEmailAndPassword(data.email, data.password)
      .then(response => {
        console.log(response.user.uid)
        if (response.user) {
          this.userData.uid = response.user.uid
          this.userData.email = response.user.email
          this.isLogin = true
          this.userData.loading = false
          this.getFarmData()
          console.log(this.userData)
          console.log(this.isLogin)
          navigation.navigate('Home')
        }
      })
      .catch(err => {
        Alert.alert(err.code, err.message)
        this.userData.loading = false
      })
  }

  firebaseSignUp = (data, navigation) => {
    console.log("Masuk sini ?")
    this.userData.loading = true
    User.createUserWithEmailAndPassword(data.email, data.password)
      .then(response => {1
        console.log('Response Daftar', response.user.uid)
        if (response.user) {
          this.userData.uid = response.user.uid
          this.userData.email = response.user.email
          this.isLogin = true
          this.userData.loading = false
          navigation.navigate('Home')
        }
        // nanti bikin db baru di firebase kalo mau ada name alamat dll
      })
      .catch(err => {
        Alert.alert(err.code, err.message)
        this.userData.loading = false
      })
  }

  firebaseSignOut = (data, navigation) => {
    // pasang alert di sini atau di component ny nanti
    User.signOut()
      .then(() => {
        this.assignUserData({uid:'',email:''})
        console.log('Logout')
      })
      .catch(err => {
        console.log(err)
        Alert.alert(err.code, err.message)
      })
  }

  triggerSiram = () => {
    console.log('Kepanggil ?')
    let userId = this.userData.uid
    let ready_siram = this.farmData.ready_siram
    if (ready_siram === 0) {
      ready_siram = 1
    } else {
      ready_siram = 0
    }

    let farmUpdate = {
      ready_siram: ready_siram,
      last_siram: firebase.database.ServerValue.TIMESTAMP,
      last_updated: firebase.database.ServerValue.TIMESTAMP
    }
    db.ref(`/farms/${userId}`).update(farmUpdate)
      .then(response => {
        console.log(response)
      })
      .catch(err => {
        console.log('Error', err)
      })
  }

  dateFormat = (times) => {
    let months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
    let date = new Date(times).getDate()
    let month = new Date(times).getMonth()
    let year = new Date(times).getFullYear()
    let minute = new Date(times).getMinutes()
    let hours = new Date(times).getHours()
    // (minute < 10)? minute = `0${minute}`: minute = `${minute}`
    // (hours < 10)? hours = `0${hours}`: hours = `${hours}`

    return (date < 10)? `0${date} ${months[month]} ${year}`: `${date} ${months[month]} ${year}, ${hours}:${minute}`
  }
}

const userStore = new UserStore() 
export default userStore