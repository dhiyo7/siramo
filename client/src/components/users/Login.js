import React, { Component } from 'react'
import { 
  View, Text, TextInput, BackHandler,
  StyleSheet, TouchableOpacity, Alert, Image, ImageBackground
} from 'react-native'
import { inject, observer } from 'mobx-react'
import Loader from '../customs/Loader'
// import { User, Fire } from '../../store/firebase'
// import SignUp from './SignUp'
import UserStore from '../../store/UserStore'

const background = require('../../assets/logo/Drawing1-ModelEdit.png')
const emailIcon = require('../../assets/logo/584856b4e0bb315b0f7675ac.png')
const passwordIcon = require('../../assets/logo/lock_512pxGREY.png')

@observer class Login extends Component {
  constructor(props) {
    super(props)
    this.state = {
      email: '',
      password: '',
      props: props
    }
    this.handleBackButtonClick = this.handleBackButtonClick.bind(this)
  }

  static navigationOptions = {
    header: null
  }

  componentWillMount() {
    BackHandler.addEventListener('hardwareBackPress', this.handleBackButtonClick)
  }

  componentWillUnmount() {
    BackHandler.removeEventListener('hardwareBackPress', this.handleBackButtonClick)
  }

  handleBackButtonClick() {
      return true
  }

  submitLogin = () => {
    let email = this.state.email
    let password = this.state.password
    
    if (!email) {
      Alert.alert('Sorry!','Please fill your Email')
    } else if (!password) {
      Alert.alert('Sorry!','Please fill your Password')
    } else {
      let navigation = this.props.navigation
      UserStore.firebaseLogin({email, password}, navigation)
    }
  }
 
  render() {
    const { loading } = UserStore.userData
    // tiap ngetik ngerender terus
    return (
      <View style={styles.mainContainer}>
        <View style={styles.container}>
          <Image
            style={styles.imageContainer}
            source={background}
          />
          <Loader loading={loading}/>
          <View style={styles.inputBox} >
            <Image
              style={styles.iconBox}         
              source={emailIcon} />
              <TextInput
                style={styles.input}
                autoCapitalize='none'
                onSubmitEditing={() => this.passwordInput.focus()}
                onChangeText={(email) => this.setState({email})}
                autoCorrect={false}
                keyboardType='email-address'
                returnKeyType='next'
                placeholder='Enter your email address'
                placeholderTextColor='#3E2723'
              />
          </View>
          <View style={styles.inputBox}> 
            <Image
              style={styles.iconBox}         
              source={passwordIcon} />
            <TextInput
              style={styles.input}
              returnKeyType='go'
              ref={(input) => this.passwordInput = input}
              onChangeText={(password) => this.setState({password})}
              placeholder='Enter your password'
              placeholderTextColor='#3E2723'
              secureTextEntry
            />
          </View>
            <TouchableOpacity
              style={styles.buttonContainer}
              onPress={this.submitLogin}
            >
              <Text style={styles.buttonText}>LOGIN</Text>
            </TouchableOpacity>

          {/* <TouchableOpacity
            style={styles.buttonContainer}
            onPress={this.googleLogin}
          >
            <Text style={styles.buttonText}>Connect with Google</Text>
          </TouchableOpacity> */}

            <TouchableOpacity
              style={styles.buttonContainer}
              onPress={() => this.props.navigation.navigate('SignUp')}
            >
              <Text style={styles.buttonText}>SIGN UP</Text>
              
            </TouchableOpacity>
        </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  mainContainer: {
    backgroundColor: '#F1F8E9',
    alignItems: 'center',
    justifyContent: 'center',
    paddingLeft: 12,
    paddingRight: 12
  },
  container: {
    backgroundColor: '#F1F8E9',
    width: '100%',
    height: '100%',
    marginRight: 8,
    marginLeft: 8,
    alignItems: 'center',
    paddingTop: 50,
  },
  imageContainer: {
    width: 200,
    height: 200,
    marginBottom: 8
  },
  inputBox: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 4,
    marginTop: 4,
    width: 400,
  },
  iconBox: {
    width: 25,
    height: 25,
    justifyContent: 'center',
    alignItems: 'center',
  },
  input: {
    height:40,
    width: 280,
    backgroundColor: 'rgba(225,225,225,0.2)',
    padding: 6,
    marginRight: 4,
    marginLeft: 4,
    marginTop: 2,
    marginBottom: 2,
    color: '#8D6E63',
    borderRadius: 4,
    fontSize: 16
  },
  buttonContainer: {
    backgroundColor: '#33691E',
    paddingVertical: 15,
    marginBottom: 8,
    borderRadius: 4,
    width: 320,
    alignItems: 'center',
    justifyContent: 'center',
  },
  buttonText: {
    color: '#fff',
    textAlign: 'center',
    fontWeight: '700'
  }
})

export default Login