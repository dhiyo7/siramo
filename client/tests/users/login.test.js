import 'react-native';
import React from 'react';
import Loader from '../../src/components/customs/Loader'
import { 
  View, Text, TextInput, BackHandler,
  StyleSheet, TouchableOpacity, Alert, Image, ImageBackground
} from 'react-native'
import Enzyme, { mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {Provider} from 'mobx-react'
import UserStore from '../../src/store/UserStore'
import Login from '../../src/components/users/Login';
Enzyme.configure({ adapter: new Adapter() })

import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const wrapper = shallow(
    <Provider UserStore={UserStore}><Login/></Provider>
  )
  expect(wrapper).toMatchSnapshot();
});

it('use all Element of react native that is imported', () => {
  const wrapper = shallow(
    <Provider UserStore={UserStore}><Login/></Provider>
  )
  expect(wrapper.containsMatchingElement([
    <View/>,
    <Text/>,
    <BackHandler/>,
    <StyleSheet/>,
    <TouchableOpacity/>,
    <Alert/>,
    <Image/>,
    <ImageBackground/>,
    <Loader/>
  ]))})

  it('button works', () => {
    const navigation = { navigate: jest.fn() }
    const wrapper = shallow(
      <Login UserStore={UserStore} navigation={navigation}/>
    )
    const render = wrapper.dive()
    render.find('TouchableOpacity').forEach(button=>{
      button.simulate('press')
    })
  })

  it('input text works', () => {
    const navigation = { navigate: jest.fn() }
    const wrapper = shallow(
      <Login UserStore={UserStore} navigation={navigation}/>
    )
    const render = wrapper.dive()
    render.find('TextInput').forEach((input,index)=>{
      input.simulate('change')
    })
  })

  

