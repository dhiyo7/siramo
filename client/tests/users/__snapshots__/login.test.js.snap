// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> snapshot renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    UserStore={
      UserStore {
        "assignUserData": [Function],
        "dateFormat": [Function],
        "firebaseLogin": [Function],
        "firebaseSignOut": [Function],
        "firebaseSignUp": [Function],
        "isLogin": false,
        "timeFormat": [Function],
        "userData": Object {
          "email": "",
          "loading": false,
          "uid": "",
        },
      }
    }
  >
    <Login />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
